contributors/design-proposals/auth/pod-restriction.md

```diff
@@ -0,0 +1,477 @@
+privileged without opening up all pods it creates to running as privileged.
+
+**Namespaced Objects**: This approach has its advantages, but it doesn't scale well to large
+organizations with many namespaces (without a system on top managing all the policies across
+namespaces). See extensions below for how we might use namespaced objects with PodRestrictions.
+
+**Bind to Pod labels**: Pod labels don't have any authorization of usage without something like
+PodRestriction, so binding to labels isn't sufficient no it's own.
+
+**Bind to Namespaces**: The approach I'm recommending. Cluster level objects use a namespace
+selector to select the namespaces it applies to. It is consistent with the authorization of create
+requests, which are authorized at the namespace level. A small number of policies can be managed at
+the cluster level, and mapped to the namespaces they should apply to.
+
+#### Optional Extensions
+
+-   Namespaced PodRestrictions - PodRestriction objects can be created in a namespace, so that
+    namespace admins can place additional restrictions within their namespace.
+
+-   Additional Bindings - The NamespaceSelector is not a top-level field so that additional binding
+    approaches may be added in the future. Examples may include: ServiceAccounts, LabelSelector,
+    NodeSelector, etc.
+
+### Policy Matching - Union or Intersection?
+
+**Intersection** refers to the model where a Pod must be valid under all PodRestrictions that
+pertain to it. The advantage of intersecting is that restrictions can be composed. For example, I
+could have a policy that frontend & backend Pods cannot share a node, and a separate policy that all
+production Pods must drop privileges. Without the ability to intersect those policies, I need 3
+different versions (either, or, both). A problem with this approach is that running a privileged
+namespace (e.g. kube-system) means that every restriction needs an exception in the
+NamespaceSelector.
+
+**Union** refers to the model where a Pod must be valid under any PodRestriction that is bound to
+it. This approach makes it easier to reason about what the restrictions on a given namespace are,
+and easier to add exceptions for namespaces, but can lead to a combinatorial explosion of different
+policies that can become unmanageable.
+
+**Proposal**: Why not both? The default behavior is intersection, but you can punch a hole through
+for a specific namespace with a union approach. An optional `BindingMode` is added to the Binding
+which specifies the behavior. A pod must be valid under either:
+
+-   ANY `BindingModeAccept` policy that is bound to it, or
+
+-   ALL `BindingModeDrop` policies that are bound to it
+
+In other words, the Pod is first checked against Accept restrictions, and is accepted if any
+match. Then, it is checked against the Drop restrictions, and rejected if any don't match. The
+default behavior at the end is to accept the pod.
```

Comment:
Alternative approach - use a firewall analogy:

- Fields in the policy are just pattern-matching, there's no concept of per-field allow vs deny
- Policy action is `allow` or `deny`. If the field matchers in a policy match the pod, the action is taken.
- Policy has a priority specifying the order. Conflicts are resolved as deny (fail closed)

Reply:
@tallclair - I'd prefer having a policy action which would be allow, deny or default

Reply:
What does a default action mean? Do you mean setting field defaults?

Reply:
yes

Reply:
FYI i've added a [#1937(comment)](https://github.com/kubernetes/community/pull/1937#issuecomment-376107468) that sums up state of policy-related standards.